cleverhans/experimental/certification/utils.py

```diff
@@ -289,7 +222,7 @@ def tf_lanczos_smallest_eigval(vector_prod_fn,
 
   # Iteration 0 of Lanczos
   q_vectors = q_vectors.write(0, w)
-  w_ = vector_prod_fn(w)
+  w_ = tf.cast(vector_prod_fn(tf.cast(w, tf.float32)), tf.float64)
```

Comment:
There are two issues with this code:

* It completely ignores dtype which is provided as an argument. So if won't work if the caller wanted to use float32 dtype. In particular this causes test failures
* It makes very specific assumptions about dtype used by vector_prod_fn, which makes it inconvenient to change code in the future

There are two obvious ways to fix it:

1. add another argument to this function `dtype_of_vector_prod_fn` and do casting using `dtype` and `dtype_of_vector_prod_fn`
2. don't do any casting in this function and instead rely on `vector_prod_fn` using the correct dtype

Approach 1 will create more clutter in the code.
Approach 2 is cleaner and more concise. Also if you consider API of `tf_lanczos_smallest_eigval` as a generic utility function then one would expect that caller of `tf_lanczos_smallest_eigval` has to provide both `dtype` and `vector_prod_fn` which is compatible with given dtype.

If you follow approach 2, then you remove any casting from here and you modify some of the functions `_m_vector_prod_fn` / `_h_vector_prod_fn` / `get_psd_product` / `get_h_product` in `dual_formulation` to do proper cast.